src/ray/object_manager/pull_manager.cc

```diff
@@ -138,31 +153,67 @@ void PullManager::PullFromRandomLocation(const ObjectID &object_id) {
                      << "already has the object. The object may have been evicted. It is "
                      << "most likely due to memory pressure, object pull has been "
                      << "requested before object location is updated.";
-    return;
+    return false;
   }
 
-  // Choose a random client to pull the object from.
-  // Generate a random index.
-  std::uniform_int_distribution<int> distribution(0, node_vector.size() - 1);
-  int node_index = distribution(gen_);
-  NodeID node_id = node_vector[node_index];
-  // If the object manager somehow ended up choosing itself, choose a different
-  // object manager.
-  if (node_id == self_node_id_) {
-    std::swap(node_vector[node_index], node_vector[node_vector.size() - 1]);
-    node_vector.pop_back();
-    RAY_LOG(WARNING)
-        << "The object manager with ID " << self_node_id_ << " is trying to pull object "
-        << object_id << " but the object table suggests that this object manager "
-        << "already has the object. It is most likely due to memory pressure, object "
-        << "pull has been requested before object location is updated.";
-    node_id = node_vector[node_index % node_vector.size()];
-    RAY_CHECK(node_id != self_node_id_);
+  NodeID node_id;
+  {
+    /*
+      A one pass algorithm for picking a uniformly random, acceptable value from a list:
+      1. If you come across an acceptable element store store it with the intention of
+      returning it later.
+      2. Next time you come across an acceptable element, replace the stored value with
+      probability 1/2.
+      n. For the nth acceptable element, store it for return with proabability 1/n.
+    */
+    int num_candidates = 0;
+    for (NodeID &potential : node_vector) {
+      if (inflight_pulls.count(potential) > 0) {
```

Comment:
The next block of code disallows node ID that matches self. Shouldn't we do that here (instead)?